storage/src/main/java/tech/pegasys/teku/storage/Store.java

```diff
@@ -354,23 +451,16 @@ public VoteTracker getVote(UnsignedLong validatorIndex) {
 
     @CheckReturnValue
     public SafeFuture<Void> commit() {
-      final StorageUpdate updateEvent =
-          new StorageUpdate(
-              genesis_time,
-              justified_checkpoint,
-              finalized_checkpoint,
-              best_justified_checkpoint,
-              blocks,
-              block_states,
-              checkpoint_states,
-              votes);
+      final TransactionCommitUpdates updates = TransactionCommitUpdates.calculate(Store.this, this);
```

Comment:
Hmm - I think the current structure is problematic.  We do these calculations upfront to figure out our updates, send those updates to storage, and only then get a `writeLock` to actually update our in-memory `Store`.  But with that setup, we could interleave multiple transactions and commit the calculated changes out of order, and run into problems while trying to calculate the updates because the `Store` can change out from under us ...

Reply:
Fixed this by adding a write lock around the update calculations.  This means that our update calculations are running off of a fixed `Store`, so we don't have to worry about any data changing while we're processing.  It's possible that we could fail to prune some new blocks that end up getting committed after we run our calculations, but those data will eventually be pruned the next time we finalize a new checkpoint.